Source/Internal/IGListPreprocessor.m

```diff
@@ -0,0 +1,98 @@
+//
+//  IGListPreprocessor.m
+//  IGListKit
+//
+//  Created by Adlai Holler on 2/20/17.
+//  Copyright © 2017 Instagram. All rights reserved.
+//
+
+#import "IGListPreprocessor.h"
+
+#import "IGListAssert.h"
+#import "IGListPreprocessingContext.h"
+#import "IGListPreprocessingDelegate.h"
+#import "IGListSectionController.h"
+#import "IGListSectionMap.h"
+
+@implementation IGListPreprocessor {
+	NSMapTable<IGListPreprocessingContext *, id<IGListPreprocessingDelegate>> *_contextToDelegateMap;
+	IGListSectionMap *_sectionMap;
+	dispatch_group_t _group;
+	dispatch_block_t _completion;
+	BOOL _ranCompletion;
+}
+
+- (instancetype)initWithSectionMap:(IGListSectionMap *)sectionMap {
+	if (self = [super init]) {
+		_contextToDelegateMap = [NSMapTable strongToStrongObjectsMapTable];
+		_sectionMap = [sectionMap copy];
+		_group = dispatch_group_create();
+	}
+	return self;
+}
+
+- (void)schedulePreprocessingWithContainerSize:(CGSize)containerSize completion:(nonnull dispatch_block_t)completionBlock {
+	IGAssertMainThread();
+	_completion = completionBlock;
```

Comment:
I haven't fully understood the scope of this preprocessor, but _completion is being overwritten here right? What happens if a second schedulePreprocessingWithContainerSize is called a second time? If so, then _completion from the first execution is overwritten, and is never executed, which could leave callers in an incomplete state permanently, right? Is calling multiple times allowed? If not, then perhaps an assertion?